Sources/NIOCore/ByteBuffer-lengthPrefix.swift

```diff
@@ -0,0 +1,107 @@
+    }
+}
+
+extension ByteBuffer {
+    /// Prefixes a message written by `writeMessage` with the number of bytes written as an `Integer`.
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to write the length prefix
+    ///     - writeMessage:
+    /// - Throws: If the number of bytes written during `writeMessage` can not be exactly represented as the given `Integer` i.e. if the number of bytes written is greater than `Integer.max`
+    /// - Returns: Number of total bytes written
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefixed<Integer>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        writeMessage: (inout ByteBuffer) throws -> Int
+    ) throws -> Int  where Integer: FixedWidthInteger {
+        var totalBytesWritten = 0
+        
+        let lengthPrefixIndex = self.writerIndex
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        totalBytesWritten += self.writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let messageLength = try writeMessage(&self)
+        totalBytesWritten += messageLength
+        
+        guard let lengthPrefix = Integer(exactly: messageLength) else {
+            throw LengthPrefixError.messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat
+        }
+        
+        self.setInteger(lengthPrefix, at: lengthPrefixIndex, endianness: endianness, as: Integer.self)
+        
+        return totalBytesWritten
+    }
+}
+
+extension ByteBuffer {
+    /// Reads an `integer` from `self` and passes it to `readMessage`. 
+    /// It is checked that exactly the number of bytes specified in the length prefix is read during the call to `readMessage`.
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to read the length prefix
+    ///     - readMessage: A closure that takes the length prefix, reads exactly that number of bytes from the given `ByteBuffer` and returns the result.
+    /// - Throws: if `readMessage` returns a non-nil value and does not read exactly the number of bytes passed to it.
+    /// - Returns: `nil` if the length prefix could not be read, 
+    ///            the length prefix is negative or
+    ///            the buffer does not contain enough bytes to read a message of this length.
+    ///            Otherwise the result of `readMessage`.
+    @inlinable
+    public mutating func readLengthPrefixed<Integer, Result>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        readMessage: (Integer, inout ByteBuffer) throws -> Result?
```

Comment:
Ok, we need to be very careful with this API, I think.

Concretely, under what circumstances should `readMessage` return `nil`? We can only call the `readMessage` function if we have an appropriate number of bytes. `ByteBuffer` traditionally uses `nil` to signal "not enough bytes to parse", but if we call `readMessage` we got as many bytes as we're ever gonna get.

I think if `readMessage` returns `nil` we need to `throw`. This helps make the `Result?` optional unambiguous: it returns `nil` only if there weren't enough bytes for the full body to be present.

Reply:
I have changed the implementation and it now no longer throws at all. Reading too much is not possible because we pass `readMessage` a `slice` now. Not consuming all bytes is still possible and does not throw. However, I think we want to actually allow this because protocols like [RTP use padding bytes at the end of a message which should be ignored](https://datatracker.ietf.org/doc/html/rfc3550#page-37).

I think it is okay to return nil if `readMessage` returns nil. This mirrors the behavior of `readString(length:)` which also returns nil if decoding fails. Otherwise the following example would throw if String decoding fails:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { length, buffer in 
    buffer.readString(length)
}
```
This would be surprising at least for me.

Reply:
`readString(length:)` does not return `nil` if decoding fails because decoding cannot fail. Specifically, we use [`String(decoding:as:)`](https://developer.apple.com/documentation/swift/string/2907004-init) which is non-failable. This is not the same as [`String(bytes:encoding:)`](https://developer.apple.com/documentation/swift/string/3126734-init) which _is_ failable.

There should be no `read` method in NIO that returns `nil` in a circumstance other than "there were not enough readable bytes to satisfy the operation", and if there is that's a bug.

Reply:
Ah I see! Thanks for taking the time to explain it to me :)